File: devpulse-backend/routers/scanning.py
from fastapi import APIRouter, Depends, HTTPException
from pydantic import BaseModel
from typing import Optional
import json

from database import (
    create_scan, update_scan, get_scans_by_collection, get_scan_by_id,
    get_findings_by_scan, update_finding_status, get_collection_by_id,
    create_finding
)
from auth import get_current_user as get_current_user_auth, User as DBUser

router = APIRouter()


class RunScanRequest(BaseModel):
    collection_id: str


class UpdateFindingRequest(BaseModel):
    status: str  # open, in_progress, resolved, false_positive


OWASP_CATEGORIES = [
    {"id": "A01", "title": "Broken Access Control", "severity": "High"},
    {"id": "A02", "title": "Cryptographic Failures", "severity": "High"},
    {"id": "A03", "title": "Injection", "severity": "Critical"},
    {"id": "A04", "title": "Insecure Design", "severity": "Medium"},
    {"id": "A05", "title": "Security Misconfiguration", "severity": "Medium"},
    {"id": "A06", "title": "Vulnerable and Outdated Components", "severity": "High"},
    {"id": "A07", "title": "Identification and Authentication Failures", "severity": "High"},
    {"id": "A08", "title": "Software and Data Integrity Failures", "severity": "Medium"},
    {"id": "A09", "title": "Security Logging and Monitoring Failures", "severity": "Low"},
    {"id": "A10", "title": "Server-Side Request Forgery (SSRF)", "severity": "Medium"},
]


def analyze_collection_for_findings(collection_data: dict, fmt: str) -> list:
    """Analyze a collection and generate security findings."""
    findings = []
    items = collection_data.get("item", []) if fmt == "postman" else []
    paths = collection_data.get("paths", {}) if fmt == "openapi" else {}

    # Check for missing auth on write endpoints
    all_endpoints = []
    if fmt == "postman":
        for item in items:
            req = item.get("request", {})
            method = (req.get("method", "GET") or "GET").upper()
            headers = req.get("header", [])
            url = req.get("url", {})
            raw_url = url.get("raw", "") if isinstance(url, dict) else str(url)
            has_auth = any(
                str(h.get("key", "")).lower() == "authorization"
                for h in headers
            )
            all_endpoints.append({"method": method, "url": raw_url, "has_auth": has_auth})
    elif fmt == "openapi":
        for path, methods in paths.items():
            for method_name, details in methods.items():
                if method_name in ("get", "post", "put", "delete", "patch"):
                    security = details.get("security", [])
                    all_endpoints.append({
                        "method": method_name.upper(),
                        "url": path,
                        "has_auth": bool(security)
                    })

    # A01: Broken Access Control
    write_endpoints_no_auth = [
        e for e in all_endpoints
        if e["method"] in ("POST", "PUT", "DELETE", "PATCH") and not e["has_auth"]
    ]
    if write_endpoints_no_auth:
        for ep in write_endpoints_no_auth[:5]:
            findings.append({
                "severity": "High",
                "category": "Broken Access Control",
                "title": f"Missing authentication on {ep['method']} {ep['url']}",
                "description": f"The endpoint {ep['method']} {ep['url']} does not require authentication.",
                "remediation": "Add authentication/authorization checks to this endpoint.",
                "endpoint": ep["url"],
            })

    # A02: Cryptographic Failures - check for HTTP URLs
    http_endpoints = [e for e in all_endpoints if e["url"].startswith("http://")]
    if http_endpoints:
        findings.append({
            "severity": "High",
            "category": "Cryptographic Failures",
            "title": "Insecure HTTP endpoints detected",
            "description": f"{len(http_endpoints)} endpoint(s) use HTTP instead of HTTPS.",
            "remediation": "Enforce HTTPS on all API endpoints.",
            "endpoint": http_endpoints[0]["url"],
        })

    # A05: Security Misconfiguration - check for risky paths
    risky_patterns = ["debug", "test", "internal", "admin", "backup", "tmp"]
    for ep in all_endpoints:
        url_lower = ep["url"].lower()
        for pattern in risky_patterns:
            if pattern in url_lower:
                findings.append({
                    "severity": "Medium",
                    "category": "Security Misconfiguration",
                    "title": f"Risky endpoint detected: {ep['url']}",
                    "description": f"Endpoint contains '{pattern}' which may expose sensitive functionality.",
                    "remediation": f"Remove or restrict access to the '{pattern}' endpoint.",
                    "endpoint": ep["url"],
                })
                break

    # A07: Auth failures - check for integer IDs (potential IDOR)
    for ep in all_endpoints:
        import re
        if re.search(r'/\d+', ep["url"]):
            findings.append({
                "severity": "Medium",
                "category": "Identification and Authentication Failures",
                "title": f"Potential IDOR vulnerability on {ep['url']}",
                "description": "Endpoint uses integer IDs which may be vulnerable to Insecure Direct Object Reference.",
                "remediation": "Use UUIDs or add authorization checks to verify resource ownership.",
                "endpoint": ep["url"],
            })
            break

    # If no findings, add a clean bill
    if not findings:
        findings.append({
            "severity": "Low",
            "category": "General",
            "title": "No critical issues detected",
            "description": "The collection passed basic security checks. Manual review recommended for deeper analysis.",
            "remediation": "Consider running a full OWASP assessment manually.",
            "endpoint": "",
        })

    return findings


@router.post("/run")
def run_scan(req: RunScanRequest, current_user: DBUser = Depends(get_current_user_auth)):
    col = get_collection_by_id(req.collection_id)
    if not col or col["user_id"] != current_user.id:
        raise HTTPException(status_code=404, detail="Collection not found or access denied")

    scan = create_scan(req.collection_id, current_user.id)
    update_scan(scan["id"], status="running")

    try:
        collection_data = json.loads(col["data"]) if isinstance(col["data"], str) else col["data"]
        raw_findings = analyze_collection_for_findings(collection_data, col["format"])

        severity_counts = {"Critical": 0, "High": 0, "Medium": 0, "Low": 0}
        for f in raw_findings:
            create_finding(
                scan_id=scan["id"],
                severity=f["severity"],
                category=f["category"],
                title=f["title"],
                description=f.get("description", ""),
                remediation=f.get("remediation", ""),
                endpoint=f.get("endpoint", ""),
            )
            severity_counts[f["severity"]] = severity_counts.get(f["severity"], 0) + 1

        update_scan(
            scan["id"],
            status="completed",
            total_findings=len(raw_findings),
            critical_count=severity_counts.get("Critical", 0),
            high_count=severity_counts.get("High", 0),
            medium_count=severity_counts.get("Medium", 0),
            low_count=severity_counts.get("Low", 0),
        )

        return {"scan_id": scan["id"], "status": "completed", "findings_count": len(raw_findings), "severity_counts": severity_counts}
    except Exception as e:
        update_scan(scan["id"], status="failed")
        raise HTTPException(status_code=500, detail=f"Scan failed: {str(e)}")


@router.get("/collection/{collection_id}")
def list_scans(collection_id: str, current_user: DBUser = Depends(get_current_user_auth)):
    col = get_collection_by_id(collection_id)
    if not col or col["user_id"] != current_user.id:
        raise HTTPException(status_code=404, detail="Collection not found")
    return {"scans": get_scans_by_collection(collection_id)}


@router.get("/{scan_id}")
def get_scan_detail(scan_id: str, current_user: DBUser = Depends(get_current_user_auth)):
    scan = get_scan_by_id(scan_id)
    if not scan or scan.get("user_id") != current_user.id:
        raise HTTPException(status_code=404, detail="Scan not found")
    findings = get_findings_by_scan(scan_id)
    return {"scan": scan, "findings": findings}


@router.patch("/findings/{finding_id}")
def patch_finding(finding_id: str, req: UpdateFindingRequest, current_user: DBUser = Depends(get_current_user_auth)):
    if req.status not in ("open", "in_progress", "resolved", "false_positive"):
        raise HTTPException(status_code=400, detail="Invalid status")
    result = update_finding_status(finding_id, req.status)
    if not result:
        raise HTTPException(status_code=404, detail="Finding not found")
    return result

File: server/utils/scanning.ts
import { nanoid } from "nanoid";

export function safeGetPath(rawUrl: string): string | null {
  if (!rawUrl) return null;
  try {
    const fullUrl = rawUrl.startsWith("http")
      ? rawUrl
      : "https://example.com" +
        (rawUrl.startsWith("/") ? rawUrl : "/" + rawUrl);
    return new URL(fullUrl).pathname;
  } catch {
    return null;
  }
}

export function generateRealFindings(collectionData: any) {
  const findings: Array<{
    id: string;
    title: string;
    severity: "Critical" | "High" | "Medium" | "Low";
    description: string;
    category: string;
    remediation: string;
    cweId: string;
  }> = [];

  const items: any[] = collectionData.item || [];
  const openApiPaths = collectionData.paths || {};

  // Process Postman-style items
  items.forEach((item: any) => {
    const url = item.request?.url?.raw || item.request?.url || "";
    const headers: any[] = item.request?.header || [];
    const method: string = (item.request?.method || "").toUpperCase();
    const displayUrl = typeof url === "string" ? url : url?.raw || "";

    // OWASP A02 - Cryptographic Failures: HTTP instead of HTTPS
    if (displayUrl.startsWith("http://")) {
      findings.push({
        id: nanoid(),
        title: "Cleartext HTTP Communication",
        severity: "High",
        description:
          "Endpoint " + displayUrl + " transmits data over unencrypted HTTP.",
        category: "Cryptographic Failures (OWASP A02)",
        remediation:
          "Enforce HTTPS on all endpoints. Set up HTTP → HTTPS redirect on your server or load balancer.",
        cweId: "CWE-319",
      });
    }

    // OWASP A07 - Authentication: mutating endpoint without auth header
    if (["POST", "PUT", "DELETE", "PATCH"].includes(method)) {
      const hasAuth = headers.some(
        (h: any) =>
          typeof h.key === "string" &&
          (h.key.toLowerCase().includes("authorization") ||
            h.key.toLowerCase().includes("x-api-key") ||
            h.key.toLowerCase().includes("api-key"))
      );
      if (!hasAuth) {
        findings.push({
          id: nanoid(),
          title: "Unauthenticated State-Changing Request",
          severity: "Critical",
          description:
            method +
            " " +
            (displayUrl || "endpoint") +
            " has no Authorization or API-Key header, making it vulnerable to unauthorized writes.",
          category: "Broken Authentication (OWASP A07)",
          remediation:
            "Add an Authorization: Bearer <token> or X-API-Key header. Validate server-side on every request.",
          cweId: "CWE-306",
        });
      }
    }

    // OWASP A01 - Broken Access Control: integer IDs in URL path (IDOR risk)
    const path = safeGetPath(displayUrl);
    if (path && /\/\d+/.test(path)) {
      findings.push({
        id: nanoid(),
        title: "Potential Insecure Direct Object Reference (IDOR)",
        severity: "Medium",
        description:
          "Endpoint " +
          path +
          " uses a sequential integer ID, which could allow unauthorized access to other users' resources.",
        category: "Broken Access Control (OWASP A01)",
        remediation:
          "Replace integer IDs with UUIDs. Always verify the authenticated user owns the resource before returning data.",
        cweId: "CWE-639",
      });
    }

    // OWASP A05 - Security Misconfiguration: debug or test headers present
    const hasDebugHeader = headers.some(
      (h: any) =>
        typeof h.key === "string" &&
        (h.key.toLowerCase().startsWith("x-debug") ||
          h.key.toLowerCase() === "x-forwarded-for")
    );
    if (hasDebugHeader) {
      findings.push({
        id: nanoid(),
        title: "Debug Headers Exposed in Request",
        severity: "Low",
        description:
          "Request to " +
          (displayUrl || "endpoint") +
          " includes debug headers that should never appear in production traffic.",
        category: "Security Misconfiguration (OWASP A05)",
        remediation:
          "Remove debug headers (X-Debug-*, X-Forwarded-For) before deploying to production.",
        cweId: "CWE-489",
      });
    }

    // OWASP A09 - Security Logging: missing request ID / correlation ID header
    if (method !== "GET") {
      const hasCorrelation = headers.some(
        (h: any) =>
          typeof h.key === "string" &&
          (h.key.toLowerCase().includes("x-request-id") ||
            h.key.toLowerCase().includes("x-correlation-id") ||
            h.key.toLowerCase().includes("request-id"))
      );
      if (!hasCorrelation) {
        findings.push({
          id: nanoid(),
          title: "Missing Request Correlation ID",
          severity: "Low",
          description:
            method +
            " " +
            (displayUrl || "endpoint") +
            " does not include a correlation/request ID header, making audit trail incomplete.",
          category: "Security Logging Failures (OWASP A09)",
          remediation:
            "Include X-Request-ID or X-Correlation-ID headers for all non-GET requests to ensure full request traceability.",
          cweId: "CWE-778",
        });
      }
    }
  });

  // Process OpenAPI paths for additional checks
  Object.entries(openApiPaths).forEach(([pathStr, pathItem]: [string, any]) => {
    Object.entries(pathItem || {}).forEach(
      ([httpMethod, operation]: [string, any]) => {
        if (["get", "post", "put", "delete", "patch"].includes(httpMethod)) {
          const hasSecurity =
            (operation.security && operation.security.length > 0) ||
            (collectionData.security && collectionData.security.length > 0);
          if (!hasSecurity && httpMethod !== "get") {
            findings.push({
              id: nanoid(),
              title: "OpenAPI Endpoint Missing Security Scheme",
              severity: "High",
              description:
                httpMethod.toUpperCase() +
                " " +
                pathStr +
                " has no security scheme defined in the OpenAPI spec.",
              category: "Broken Authentication (OWASP A07)",
              remediation:
                "Add a security: [] block to this operation referencing your securitySchemes (e.g. bearerAuth).",
              cweId: "CWE-306",
            });
          }
        }
      }
    );
  });

  return findings;
}

export function detectShadowAPIs(collectionData: any) {
  const shadowAPIs: Array<{
    endpoint: string;
    method: string;
    riskLevel: "LOW" | "MEDIUM" | "HIGH" | "CRITICAL";
    reason: string;
    recommendation: string;
  }> = [];

  const items: any[] = collectionData.item || [];
  const riskyKeywords = [
    "debug",
    "test",
    "internal",
    "admin",
    "hidden",
    "dev",
    "beta",
    "staging",
    "old",
    "backup",
    "temp",
    "tmp",
    "legacy",
  ];
  const safePrefixes = [
    "/api/v1",
    "/api/v2",
    "/api/v3",
    "/public",
    "/v1",
    "/v2",
    "/v3",
  ];
  const seen = new Set<string>();

  items.forEach((item: any) => {
    const method = (item.request?.method || "GET").toUpperCase();
    const rawUrl = item.request?.url?.raw || item.request?.url || "";
    const displayUrl = typeof rawUrl === "string" ? rawUrl : rawUrl?.raw || "";
    const path = safeGetPath(displayUrl) || "/";

    const key = method + ":" + path;
    if (seen.has(key)) return;
    seen.add(key);

    const lowerPath = path.toLowerCase();
    const matchedKeyword = riskyKeywords.find(k => lowerPath.includes(k));
    const isUnusualPath = !safePrefixes.some(prefix =>
      lowerPath.startsWith(prefix)
    );
    const isDestructiveWithoutPrefix =
      (method === "DELETE" || method === "PUT") &&
      !safePrefixes.some(p => lowerPath.startsWith(p));
    const isUndocumentedAdminOp =
      (lowerPath.includes("admin") || lowerPath.includes("internal")) &&
      ["DELETE", "PUT", "POST"].includes(method);

    let riskLevel: "LOW" | "MEDIUM" | "HIGH" | "CRITICAL" = "LOW";
    let reason = "";
    let recommendation = "";

    if (isUndocumentedAdminOp) {
      riskLevel = "CRITICAL";
      reason =
        "Admin/internal endpoint detected with destructive " +
        method +
        " operation";
      recommendation =
        "Document all admin endpoints in your API spec and restrict access via authentication/authorization.";
    } else if (isDestructiveWithoutPrefix) {
      riskLevel = "HIGH";
      reason =
        "Destructive " +
        method +
        " endpoint without standard API versioning prefix";
      recommendation =
        "Add API versioning prefix (/api/v1) and document this endpoint.";
    } else if (matchedKeyword) {
      riskLevel = "MEDIUM";
      reason = "Endpoint path contains risky keyword: " + matchedKeyword;
      recommendation =
        "Review if " +
        matchedKeyword +
        " endpoint should be publicly accessible or documented.";
    } else if (isUnusualPath) {
      riskLevel = "LOW";
      reason = "Endpoint uses non-standard path pattern";
      recommendation =
        "Consider using standard REST conventions with API versioning.";
    }

    if (
      matchedKeyword ||
      isUnusualPath ||
      isDestructiveWithoutPrefix ||
      isUndocumentedAdminOp
    ) {
      shadowAPIs.push({
        endpoint: path,
        method,
        riskLevel,
        reason,
        recommendation,
      });
    }
  });

  return shadowAPIs;
}

export function calculateRiskScore(findings: any[]) {
  let score = 0;
  for (const finding of findings) {
    if (finding.severity === "Critical") score += 30;
    else if (finding.severity === "High") score += 20;
    else if (finding.severity === "Medium") score += 10;
    else if (finding.severity === "Low") score += 5;
  }
  return Math.min(100, score);
}

export function getRiskLevel(
  score: number
): "LOW" | "MEDIUM" | "HIGH" | "CRITICAL" {
  if (score >= 80) return "CRITICAL";
  if (score >= 60) return "HIGH";
  if (score >= 30) return "MEDIUM";
  return "LOW";
}

export function generatePCIDSSRequirements(collectionData: any) {
  const items: any[] = collectionData.item || [];
  const paths = collectionData.paths || {};
  const allItems = [
    ...items.map((i: any) => ({
      url: i.request?.url?.raw || i.request?.url || "",
      method: (i.request?.method || "GET").toUpperCase(),
      headers: i.request?.header || [],
    })),
    ...Object.entries(paths).flatMap(([path, methods]: [string, any]) =>
      Object.entries(methods || {})
        .filter(([m]) => ["get", "post", "put", "delete", "patch"].includes(m))
        .map(([method, op]: [string, any]) => ({
          url: path,
          method: method.toUpperCase(),
          headers: [],
          hasSecurity: !!(
            op?.security?.length || collectionData.security?.length
          ),
        }))
    ),
  ];

  const requirements = [];

  // Req 1: TLS for all connections
  const hasHttp = allItems.some((i: any) => {
    const url = typeof i.url === "string" ? i.url : "";
    return url.startsWith("http://");
  });
  requirements.push({
    id: "PCI-1.1",
    title: "TLS for All Connections",
    description: "All API endpoints must use HTTPS/TLS for data transmission.",
    status: hasHttp ? "not_met" : "met",
  });

  // Req 2: Authentication for sensitive operations
  const writeOpsWithoutAuth = allItems.filter(
    (i: any) =>
      ["POST", "PUT", "DELETE", "PATCH"].includes(i.method) &&
      !i.headers.some(
        (h: any) =>
          h.key?.toLowerCase().includes("authorization") ||
          h.key?.toLowerCase().includes("api-key")
      )
  );
  requirements.push({
    id: "PCI-2.1",
    title: "Authentication for Write Operations",
    description: "All POST/PUT/DELETE operations must require authentication.",
    status: writeOpsWithoutAuth.length > 0 ? "not_met" : "met",
  });

  // Req 3: No debug endpoints in production
  const hasDebug = allItems.some((i: any) => {
    const url = (typeof i.url === "string" ? i.url : "").toLowerCase();
    return (
      url.includes("debug") || url.includes("test") || url.includes("internal")
    );
  });
  requirements.push({
    id: "PCI-3.1",
    title: "No Debug/Test Endpoints",
    description:
      "Debug, test, and internal endpoints must not be exposed in production.",
    status: hasDebug ? "not_met" : "met",
  });

  // Req 4: Input validation (check for obvious injection patterns)
  const hasQueryParams = allItems.some((i: any) => {
    const url = typeof i.url === "string" ? i.url : "";
    return url.includes("?") || url.includes("{");
  });
  requirements.push({
    id: "PCI-4.1",
    title: "Input Validation",
    description: "All user input must be validated before processing.",
    status: hasQueryParams ? "manual_review" : "met",
  });

  // Req 5: Audit logging
  requirements.push({
    id: "PCI-5.1",
    title: "Audit Logging",
    description: "All access to cardholder data must be logged.",
    status: "manual_review",
  });

  return requirements;
}

export function generateOWASPRequirements(collectionData: any) {
  const items: any[] = collectionData.item || [];
  const paths = collectionData.paths || {};
  const allItems = [
    ...items.map((i: any) => ({
      url: i.request?.url?.raw || i.request?.url || "",
      method: (i.request?.method || "GET").toUpperCase(),
      headers: i.request?.header || [],
    })),
    ...Object.entries(paths).flatMap(([path, methods]: [string, any]) =>
      Object.entries(methods || {})
        .filter(([m]) => ["get", "post", "put", "delete", "patch"].includes(m))
        .map(([method, _op]: [string, any]) => ({
          url: path,
          method: method.toUpperCase(),
          headers: [],
        }))
    ),
  ];

  return [
    {
      id: "OWASP-A01",
      title: "Broken Access Control",
      description:
        "Access control checks should be enforced server-side for every request.",
      status: allItems.some((i: any) => /\/\d+/.test(i.url))
        ? "manual_review"
        : "met",
    },
    {
      id: "OWASP-A02",
      title: "Cryptographic Failures",
      description:
        "All data transmission should use strong encryption (TLS 1.2+).",
      status: allItems.some((i: any) =>
        (typeof i.url === "string" ? i.url : "").startsWith("http://")
      )
        ? "not_met"
        : "met",
    },
    {
      id: "OWASP-A03",
      title: "Injection",
      description:
        "User-supplied data should be validated, sanitized, and escaped.",
      status: "manual_review",
    },
    {
      id: "OWASP-A04",
      title: "Insecure Design",
      description: "Security should be integrated into the design phase.",
      status: "manual_review",
    },
    {
      id: "OWASP-A05",
      title: "Security Misconfiguration",
      description:
        "Systems should be hardened with minimal features and secure defaults.",
      status: allItems.some((i: any) =>
        (typeof i.url === "string" ? i.url : "").toLowerCase().includes("debug")
      )
        ? "not_met"
        : "met",
    },
    {
      id: "OWASP-A06",
      title: "Vulnerable and Outdated Components",
      description:
        "Components should be kept up to date with known vulnerabilities patched.",
      status: "manual_review",
    },
    {
      id: "OWASP-A07",
      title: "Identification and Authentication Failures",
      description:
        "Authentication should be implemented correctly using secure mechanisms.",
      status: allItems.some(
        (i: any) =>
          ["POST", "PUT", "DELETE"].includes(i.method) &&
          !i.headers.some((h: any) =>
            h.key?.toLowerCase().includes("authorization")
          )
      )
        ? "not_met"
        : "met",
    },
    {
      id: "OWASP-A08",
      title: "Software and Data Integrity Failures",
      description:
        "Software updates and critical data should be verified for integrity.",
      status: "manual_review",
    },
    {
      id: "OWASP-A09",
      title: "Security Logging and Monitoring Failures",
      description: "Security events should be logged and monitored.",
      status: "manual_review",
    },
    {
      id: "OWASP-A10",
      title: "Server-Side Request Forgery (SSRF)",
      description: "Server-side requests should be validated and restricted.",
      status: "manual_review",
    },
  ];
}
